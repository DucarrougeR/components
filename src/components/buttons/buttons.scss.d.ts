export const btn: string;
export const btnDanger: string;
