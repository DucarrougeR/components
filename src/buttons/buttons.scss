$button-default-color: #64518A;
$button-default-highlight: #473764;
$button-light-color: transparent;
$button-scary-color: #E1332C;

$button-font-color: white;
$button-font-color-light: $button-default-color;

$button-font-size-xs: 12px;
$button-font-size-small: 14px;
$button-font-size-medium: 16px;
$button-font-size-large: 22px;

.button {
  background-color: $button-default-color;
  border: 1px solid $button-default-color;
  border-radius: 6px;
  color: $button-font-color;
  cursor: pointer;
  display: inline-block;
  font-size: $button-font-size-medium;
  line-height: 24px;
  padding: 6px 16px;
  white-space: nowrap;
  vertical-align: middle;

  &:hover {
    background-color: $button-default-highlight;
    border: 1px solid $button-default-highlight;
  }

  &.size-extra-small {
    font-size: $button-font-size-xs;
    line-height: 20px;
    padding: 2px 8px;
  }

  &.size-small {
    font-size: $button-font-size-small;
    padding: 3px 14px;
  }

  &.size-large {
    font-size: $button-font-size-large;
    line-height: 30px;
    padding: 8px 24px;
  }

  &.mode-light {
    background: $button-light-color;
    color: $button-font-color-light;
  }

  &.mode-ghost {
    background: $button-light-color;
    border: none;
    color: $button-font-color-light;
  }

  &.mode-scary {
    background: $button-scary-color;
    border-color: $button-scary-color;
  }
}
