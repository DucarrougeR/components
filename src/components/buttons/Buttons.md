# Buttons

```js
const grid = require('./button-modes').grid;
grid(['ghost', 'scary'], ['small', 'large']);
```

```js

let inputVal = '';
let changer = function(event) {
  inputVal = event.target.value
};

<div>
  <input name="foo" onChange={changer} />
  <Button mode={inputVal}>Hello World</Button>
  <Button mode="ghost">Hello World</Button>
  <Button mode="scary">Hello World</Button>
</div>
```
