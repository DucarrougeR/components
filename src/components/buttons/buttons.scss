@import '../../styles/font-utils';

$light-normal-button: #FFFFFF;
$light-hover-button: #EAE6FA;
$light-active-button: #D6CDF5;
$light-normal-button-border: #7E64E0;
$light-hover-button-border: #4C33AA;
$light-active-button-border: #4C33AA;

$scary-normal-button: #E1332C;
$scary-hover-button: #B00505;
$scary-active-button: #92100E;

$disabled-grey-button: #E4E5E6;

$text-white: #FFFFFF;
$text-grey: #A4A6A9;
$text-purple: #7E64E0;

.button {
  @include fontSize(16px);
  background-color: var(--lens-primary-color);
  border: pxToRem(1px) solid var(--lens-primary-color-border);
  border-radius: pxToRem(6px);
  color: var(--lens-primary-color-text);
  cursor: pointer;
  display: inline-block;
  line-height: pxToRem(24px);
  padding: pxToRem(6px) pxToRem(16px);

  white-space: nowrap;
  vertical-align: middle;

  &:hover, &.hover {
    background-color:  var(--lens-primary-color-hover);
    border-color: var(--lens-primary-color-hover);
  }

  &:active, &.active {
    background-color: var(--lens-primary-color-active);
    border-color: var(--lens-primary-color-active);
  }

  &.size-extra-small {
    @include fontSize(12px);
    line-height: pxToRem(20px);
    padding: pxToRem(2px) pxToRem(8px);
  }

  &.size-small {
    @include fontSize(14px);
    padding: pxToRem(3px) pxToRem(14px);
  }

  &.size-large {
    @include fontSize(22px);
    line-height: pxToRem(30px);
    padding: pxToRem(8px) pxToRem(24px);
  }

  &.mode-light {
    background: $light-normal-button;
    color: $text-purple;
    border-color: $light-normal-button-border;

    &:hover {
      background-color: $light-hover-button;
      border-color: $light-hover-button-border;
    }

    &:active {
      background-color: $light-active-button;
      border-color: $light-active-button-border;
    }
  }

  &.mode-ghost {
    background: $light-normal-button;
    border-color: transparent;
    color: $text-purple;

    &:hover {
      background-color: $light-hover-button;
    }

    &:active {
      background-color: $light-active-button;
    }
  }

  &.mode-scary {
    background: $scary-normal-button;
    border-color: $scary-normal-button;

    &:hover {
      border-color: $scary-hover-button;
      background-color: $scary-hover-button;
    }

    &:active {
      border-color: $scary-active-button;
      background-color: $scary-active-button;
    }
  }

  &.mode-disabled {
    background: $disabled-grey-button;
    border-color: $disabled-grey-button;
    color: #A4A6A9;
  }
}
