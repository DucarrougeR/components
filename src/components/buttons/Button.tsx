import * as React from 'react'
import * as classNames from 'classnames'
import * as styles from './buttons.scss'
import { Button as SButton } from 'semantic-ui-react'

interface ButtonProps {
  className?: string
  danger?: boolean
  onClick?: (...args: any[]) => void
}

export const Button: React.StatelessComponent<ButtonProps> = ({children, className, danger, onClick}) => {
  const btnClass = classNames(
    styles.btn,
    className,
    {
      [styles.btnDanger]: danger
    })

  return <SButton className={btnClass} onClick={onClick ? onClick : () => {}}>{children}</SButton>
}

