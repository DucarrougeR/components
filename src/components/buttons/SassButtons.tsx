import * as React from 'react'
import * as classNames from 'classnames'
import * as styles from './buttons.scss'

interface ButtonProps {
  className?: string
  danger?: boolean
  onClick?: (...args: any[]) => void
}

const Button: React.StatelessComponent<ButtonProps> = ({children, className, danger, onClick}) => {
  const btnClass = classNames(styles.btn, className, {
    [styles.btnDanger]: danger
  })

  return <button className={btnClass} onClick={onClick ? onClick : () => {}}>{children}</button>
}

export default Button
