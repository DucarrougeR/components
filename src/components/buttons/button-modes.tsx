import * as React from 'react'
import { Button, LookerButtonHTMLAttributes } from './buttons'

export function grid (modes: any[], sizes: any[]) {
  modes.push(undefined)
  sizes.push(undefined)
  const matrix: React.ReactElement<LookerButtonHTMLAttributes>[][] = []


  modes.forEach((mode) => {
    const buttonRow: React.ReactElement<LookerButtonHTMLAttributes>[] = []
    sizes.forEach((size) => {
      let desc = ''
      switch (size) {
        case 'extraSmall':
          desc = 'X-Small'
          break
        case 'small':
          desc = 'Small'
          break
        case 'large':
          desc = 'Large'
          break
        default:
          desc = 'Normal'
          break
      }
      buttonRow.push(<Button size={size} mode={mode}>{desc}</Button>)
    })
    matrix.push(buttonRow)
  })

  return (
    <div>
      <table className="fizzbuzz">
        <tbody>
        {matrix.map((row, idx) => {
          return (
            <tr key={idx}>
              {row.map((button, idx) => (<td key={idx}>{button}</td>))}
            </tr>
          )
        })}
        </tbody>
      </table>
    </div>
  )
}
