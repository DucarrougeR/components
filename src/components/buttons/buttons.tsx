import * as React from 'react'
import * as classNames from 'classnames'
import * as styles from './buttons.scss'
import { capitalize } from '../../utils/strings'

export interface LookerButtonHTMLAttributes extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  [key: string]: any
  /** mode is the color style */
  mode?: string | undefined
  size?: string | undefined
  state?: 'active' | 'hover' | undefined
}

export {
  styles as buttonStyles
}

type PropertyBag = {
  [key: string]: any
}

/**
 * Here's a basic button
 * @param {string} className
 * @param {string | undefined} size
 * @param {"default" | "ghost" | undefined} mode
 * @param {*} args
 * @returns {any}
 * @constructor
 */
export const Button = ({className, size, mode, state, ...args}: LookerButtonHTMLAttributes) => {
  const styleableProps: PropertyBag = {
    [styles.active]: state == 'active',
    [styles.hover]: state == 'hover'
  }
  if (size) styleableProps[(styles as PropertyBag)[`size${capitalize(size)}`]] = true
  if (mode) styleableProps[(styles as PropertyBag)[`mode${capitalize(mode)}`]] = true
  if (args.disabled) styleableProps['modeDisabled'] = true

  return (
    <button className={classNames(styles.button, className, styleableProps)} {...args}>{args.children}</button>
  )
}
