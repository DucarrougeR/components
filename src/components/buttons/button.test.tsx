import * as React from 'react'
import { Button } from './buttons'
import { create } from 'react-test-renderer'

test('Button states', () => {
  const component = create(<Button>ğŸ¥‘</Button>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Button ghost', () => {
  const component = create(<Button mode="ghost">ğŸ¥‘</Button>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
