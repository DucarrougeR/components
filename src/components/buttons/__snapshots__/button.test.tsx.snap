// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button ghost 1`] = `
<button
  className=""
>
  ðŸ¥‘
</button>
`;

exports[`Button states 1`] = `
<button
  className=""
>
  ðŸ¥‘
</button>
`;
